core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -140,20 +195,13 @@ private class PartitionsScan extends StaticTableScan {
   }
 
   static class PartitionMap {
-    private final Map<StructLikeWrapper, Partition> partitions = Maps.newHashMap();
-    private final Types.StructType type;
-    private final StructLikeWrapper reused;
-
-    PartitionMap(Types.StructType type) {
-      this.type = type;
-      this.reused = StructLikeWrapper.forType(type);
-    }
+    private final Map<StructLike, Partition> partitions = Maps.newHashMap();
 
     Partition get(StructLike key) {
-      Partition partition = partitions.get(reused.set(key));
+      Partition partition = partitions.get(key);
       if (partition == null) {
         partition = new Partition(key);
-        partitions.put(StructLikeWrapper.forType(type).set(key), partition);
```

Comment:
I'm a bit confused by this change, I believe the issue here is that StructType does not have a well defined hashFunction (since implementations can do whatever they like) which is why we use the Wrapper to make sure we have a valid hash.

Reply:
Changed to map of PartitionData (I feel, it should have been that way in the beginning)